ux: WalletConnect login

tomholford opened the pull request:
Closes #601

# Changes

- adds [WalletConnect](https://walletconnect.org/) login flow

# Preview

## Ready to Connect

![image](https://user-images.githubusercontent.com/16504501/127941380-ccb2df1c-7869-4673-8df9-47d13ced48b6.png)

## Connected, ready for Auth

![image](https://user-images.githubusercontent.com/16504501/128973379-5cb40f7c-52ef-430a-9531-4d000e45eb4c.png)

## Login Flow

This was tested using the [WalletConnect online test wallet](https://test.walletconnect.org/)

![scaledOutput](https://user-images.githubusercontent.com/16504501/127820120-8a590dd1-514e-4fd3-9ed8-8ee8d3816692.gif)

# Testing

To verify functionality, I downloaded several apps for both iOS and Android. After some trial and error, I realized that many wallet apps automatically reject Signing Transactions if the connected wallet has a zero balance. Once I connected a main net wallet with some funds, I was able to confirm that both Trust wallets for iOS and Android (in addition to the reference implementation in the Test Wallet) are working as expected. As for the remaining unknown wallets, I stopped testing once I found 3 working peer implementations so we can get this out the door and focus on Roller.

In summary, confirming that this works across several WalletConnect peer implementations (reference Test Wallet, Trust, and TokenPocket). Thus, if other wallets do have issues once we ship to production, we should encourage affected users to file an issue with vendor's repo.

![image](https://user-images.githubusercontent.com/16504501/129429905-edc6c123-9ca6-4b2b-9184-60b569bc0e08.png)

tomholford commented:
@Fang- A few items for when you have a moment:

-  When initially loading the WC Login component, some event listeners are instantiated to handle incoming events from the peer device. After the User authenticates and is redirected to the Home view, should the application re-instantiate WC event listeners? If it does, then if the User disconnects on their WC peer, Bridge could log the user out as well. However, this might be unexpected behavior, as it might violate the principal of least surprise. Would a user expect to be auto-logged out of Bridge by disconnecting on the peer device?
- Have tested this so far with WC Test Web Wallet, and both Coinomi and Rainbow for iOS. The test wallet and Rainbow are pretty reliable, but Coinomi would crash on occasion when authenticating. Do you have any devices with which you could also test the flow? Would be helpful to collect additional data points
- The remaining piece - allowing the user to set a custom HD Path - is dependent on using the `BridgeForm` components built-in behaviors (such as `onSubmit`); currently the implementation is hacking around it. Got stuck when I was trying to figure it out the other day, will take another crack at it today. Depending on how things go, I might be reaching out again :)

tomholford commented:
@jyng If you have a moment, please LMK if you have any thoughts or feedback, thanks

jyng commented:
1) We should use the same button type as the other options.
2) I seem to only be able to click "Connect" once. If I close it and try to click "Connect" again, nothing happens. No log in console either.

tomholford commented:
> 1. We should use the same button type as the other options.
Got it, will make that change 👍 

> 2. I seem to only be able to click "Connect" once. If I close it and try to click "Connect" again, nothing happens. No log in console either.
Okay thanks for sharing this, will investigate to see if I can reproduce the issue. What browser and WalletConnect peer are you using?

tomholford commented:
@jyng Addressed both issues, thanks for your feedback! Please LMK if you have any other thoughts, thanks

![image](https://user-images.githubusercontent.com/16504501/127941430-6e76fc31-186a-4dd7-9192-b5e16ce83c46.png)

jyng commented:
> Connect a peer that support WalletConnect
Is 'peer' a term that WalletConnect uses? If not, I think 'wallet' should replace it.

Fang- commented on src/lib/useWalletConnect.ts:
```diff
@@ -0,0 +1,202 @@
+      walletType: WALLET_TYPES.WALLET_CONNECT,
+    });
+
+    // Inspired by the MetamaskWallet implementation :)
+    // TODO: We should refactor / unify how we handle wallets
+    const wallet: WalletConnectWallet = {
+      address,
+    };
```
I mean, we do `setWallet` with a consistent type, don't we? But there is some necessary inconsistency between wallets when it comes to tx signing/sending and such.

Fang- commented on src/lib/useWalletConnect.ts:
```diff
@@ -0,0 +1,202 @@
+
+    setAuthToken(Just(authToken));
+    setWallet(Just(wallet));
+    setWalletHdPath(hdPath);
```
Does this ever come into play anywhere? It looks like in the WalletConnect case, we just connect to some specific account, but don't know anything about its origin/corresponding path. Might be misleading to set it/provide a login input for it.  
I think we only really use it in hardware wallet scenarios, where we need to tell the device to sign with a specific account?

tomholford replied:
Good call, removed this 👍

Fang- reviewed (changes requested):
When I try to generate & sign a (planet acceptance) transaction, I get `can't access property Symbol.iterator, source is undefined`. This is with a WalletConnect test wallet.

As mentioned out of band, I can't get this to work with Ledger Live (it opens the app, but the app doesn't actually do anything). May or may not be something we have control over.

galenwp commented:
very happy to see walletconnect going in!

Fang- commented on src/lib/useEthereumTransaction.js:
```diff
@@ -119,13 +128,15 @@ export default function useEthereumTransaction(
       setSignedTransactions(txns);
       setState(STATE.SIGNED);
     } catch (error) {
+      console.error(error);
```
```suggestion
      console.error('error signing txs', error);
```

We might not always be able to see a legible trace of where the error came from, so labeling it seems nice.

Fang- reviewed (commented):
Changes herein all look good. I'll want to go and test all the various login/signing methods to make sure no breakage is sneaking by, but otherwise this is in a good state.

What we should probably do, considering the flakiness of some WalletConnect clients, is add a tiny little disclaimer in its login box, saying something like "WalletConnect is a young protocol, not all wallets may work fully" or something along those lines, to give our "yes we know" responses to the inevitable "my wallet doesn't connect!" reports a little more oomph.

tomholford commented:
> Changes herein all look good. I'll want to go and test all the various login/signing methods to make sure no breakage is sneaking by, but otherwise this is in a good state.
> 
> What we should probably do, considering the flakiness of some WalletConnect clients, is add a tiny little disclaimer in its login box, saying something like "WalletConnect is a young protocol, not all wallets may work fully" or something along those lines, to give our "yes we know" responses to the inevitable "my wallet doesn't connect!" reports a little more oomph.

Good thinking. Since the action is obvious (click the big Connect button), replaced the previous help text with the disclaimer:

![image](https://user-images.githubusercontent.com/16504501/129430353-23f865c0-297b-4244-bfc7-12178f8333ee.png)


Also @Fang- , added a *Testing* section to the PR description above. In summary:
>Confirming that this works across several WalletConnect peer implementations (reference Test Wallet, Trust, and TokenPocket).

Feeling pretty comfortable about this now, will re-request review 🔁

Fang- reviewed (approved):
Alright, I think this is good for release. Let's do this!

urcades commented:
this is huge, thanks so much @tomholford
